aten/src/ATen/native/Memory.cpp

```diff
@@ -1,18 +1,49 @@
 #include <ATen/ATen.h>
 #include <ATen/MemoryOverlap.h>
 #include <ATen/NativeFunctions.h>
+#include <ATen/TensorUtils.h>
 #include <ATen/detail/CUDAHooksInterface.h>
 #include <c10/util/Exception.h>
 #include <c10/core/Storage.h>
-#include <ATen/TensorUtils.h>
 
 namespace at {
 namespace native {
 
+namespace {
+static int64_t inline get_device_index_with_primary_context() {
+  const auto& cuda_hooks = detail::getCUDAHooks();
+  // check current device first
+  int64_t current_device_index = cuda_hooks.current_device();
+  if (current_device_index >= 0) {
+    if (cuda_hooks.hasPrimaryContext(current_device_index)) {
+      return current_device_index;
+    }
+  }
+  for (int64_t device_index = 0; device_index < cuda_hooks.getNumGPUs(); device_index++) {
+    if (device_index == current_device_index) continue;
+    if (cuda_hooks.hasPrimaryContext(device_index)) {
+      return device_index;
+    }
+  }
+  return -1;
+}
+}
+
 Tensor pin_memory(const Tensor& self) {
   if (self.type().backend() != Backend::CPU) {
     AT_ERROR("cannot pin '", self.type().toString(), "' only dense CPU tensors can be pinned");
   }
+
+  // Pinned memory pointers allocated by any device can be directly used by any
+  // other device, regardless of the current device at the time of allocation,
+  // since we assume unified addressing.
+  // So we grab any existing primary context, if available.
+  // See pytorch/pytorch#21081.
+  at::OptionalDeviceGuard device_guard;
+  auto primary_ctx_device_index = get_device_index_with_primary_context();
+  if (primary_ctx_device_index >= 0) {
+    device_guard.reset_device(at::Device(at::DeviceType::CUDA, primary_ctx_device_index));
+  }
```

Comment:
Same needed in the: `Tensor empty_cpu(IntArrayRef size, const TensorOptions& options, c10::optional<c10::MemoryFormat> optional_memory_format)`

Or you can shove it into the getPinnedMemoryAllocator?

Reply:
A device guard `getPinnedMemoryAllocator` will be deallocated when the function returns, so it would be useless. Instead, I moved it to `THCCachingHostAllocator`'s `alloc`, which should be fine :)